src/mapstorage/XmlMapStorage.cpp

```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
     qWarning() << "XmlMapStorage does not implement newData()";
 }
 
-bool XmlMapStorage::loadData()
+bool XmlMapStorage::mergeData()
 {
     return false;
 }
 
-bool XmlMapStorage::mergeData()
+// ---------------------------- XmlMapStorage::loadData() ----------------------
+bool XmlMapStorage::loadData()
 {
-    return false;
+    // clear previous map
+    m_mapData.clear();
+    try {
+        log("Loading data ...");
+        QXmlStreamReader stream(m_file);
+        loadWorld(stream);
+        log("Finished loading.");
+
+        m_mapData.checkSize();
+        m_mapData.setFileName(QString(), false);
+
+        emit sig_onDataLoaded();
+        return true;
+    } catch (const std::exception &ex) {
+        const auto msg = QString::asprintf("Exception: %s", ex.what());
+        log(msg);
+        qWarning().noquote() << msg;
+
+        QMessageBox::critical(checked_dynamic_downcast<QWidget *>(parent()),
+                              tr("XmlMapStorage Error"),
+                              msg);
+
+        m_mapData.clear();
+        return false;
+    }
+}
+
+void XmlMapStorage::loadWorld(QXmlStreamReader &stream)
+{
+    ProgressCounter &progressCounter = getProgressCounter();
+    progressCounter.reset();
+    progressCounter.increaseTotalStepsBy(32728); // just a guess
```

Comment:
We could add a room count to the XML file, as long as it's only used to set the progress bar range:
I don't want to require an exact count, because it would force users to count the rooms after they edited a file - a completely unnecessary chore.

But since it's only for the progress bar, maybe we could also live without the room count?

Reply:
[UPDATE] I found a better solution:
retrieve file size when opening the XML file, and divide `QXmlStreamReader::characterOffset()` by file size to compute the progress. I will implement it.

Reply:
Done.